src/prpy/tsr/tsr.py

```diff
@@ -231,14 +334,113 @@ def from_yaml(x, *args, **kw_args):
+        """
+
+        if len(xyzrpy) != len(self.TSRs):
+            raise('Sample must be of equal length to TSR chain!')
+
+        for idx in range(len(self.TSRs)):
+            if not self.TSRs[idx].is_valid(xyzrpy[idx], ignoreNAN):
+                return False
+
+        return True
+
+    def to_transform(self, xyzrpy):
+        """
+        Converts a xyzrpy list into an
+        end-effector transform.
+
+        @param  a list of xyzrpy values
+        @return trans 4x4 transform
+        """
 
+        if not self.is_valid(xyzrpy):
+            raise ValueError('Invalid xyzrpy')
         T0_w = self.TSRs[0].T0_w
         for idx in range(len(self.TSRs)):
             tsr_current = self.TSRs[idx]
             tsr_current.T0_w = T0_w
-            T0_w = tsr_current.sample()
+            T0_w = tsr_current.to_transform(xyzrpy[idx])
 
         return T0_w
+
+    def sample_xyzrpy(self, xyzrpy=None):
+        """
+        Samples from Bw to generate an xyzrpy sample
+        Can specify some values optionally as NaN.
+
+        @param xyzrpy   (optional) a list of Bw with float('nan') for
+                        dimensions to sample uniformly.
+        @return sample  a list of sampled xyzrpy
+        """
+
+        if xyzrpy is None:
+            xyzrpy = [NANBW]*len(self.TSRs)
+
+        sample = []
+        for idx in range(len(self.TSRs)):
+            sample.append(self.TSRs[idx].sample_xyzrpy(xyzrpy[idx]))
+
+        return sample
+
+    def sample(self, xyzrpy=None):
+        """
+        Samples from the Bw chain to generate an end-effector transform.
+        Can specify some Bw values optionally.
+
+        @param xyzrpy   (optional) a list of xyzrpy with float('nan') for
+                        dimensions to sample uniformly.
+        @return T0_w    4x4 transform
+        """
+        return self.to_transform(self.sample_xyzrpy(xyzrpy))
+
+    def distance(self, trans):
+        """
+        Computes the Geodesic Distance from the TSR chain to a transform
+        @param trans 4x4 transform
+        @return dist Geodesic distance to TSR
+        @return bwopt Closest Bw value to trans output as a list of xyzrpy
+        """
+        import scipy.optimize
+
+        def objective(xyzrpy):
+            bw_stack = xyzrpy.reshape(len(self.TSRs), 6)
+            bwtrans = self.to_transform(bw_stack)
+            return prpy.util.GeodesicDistance(bwtrans, trans)
+
+        bwinit = []
+        bwbounds = []
+        for idx in range(len(self.TSRs)):
+            Bw = self.TSRs[idx].Bw
+            bwinit.extend((Bw[:, 0] + Bw[:, 1])/2)
+            bwbounds.extend([(Bw[i, 0], Bw[i, 1]) for i in range(6)])
+
+        bwopt, dist, info = scipy.optimize.fmin_l_bfgs_b(
+                                objective, bwinit, fprime=None,
+                                args=(),
+                                bounds=bwbounds, approx_grad=True)
+        return dist, bwopt.reshape(len(self.TSRs), 6)
```

Comment:
This looks good, I would just rename `xyzrpy` as `xyzrpy_stacked` or something like that.